src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs

```diff
@@ -16,9 +17,89 @@ public class ResourceNotificationService(ILogger<ResourceNotificationService> lo
+    /// </summary>
+    /// <param name="logger">The logger.</param>
+    /// <param name="hostApplicationLifetime">The host application lifetime.</param>
+    public ResourceNotificationService(ILogger<ResourceNotificationService> logger, IHostApplicationLifetime hostApplicationLifetime)
+        : this(logger)
+    {
+        ArgumentNullException.ThrowIfNull(hostApplicationLifetime);
+
+        _applicationStopping = hostApplicationLifetime.ApplicationStopping;
+    }
+
+    /// <summary>
+    /// Waits for a resource to reach the specified state. See <see cref="KnownResourceStates"/> for common states.
+    /// </summary>
+    /// <remarks>
+    /// This method returns a task that will complete when the resource reaches the specified target state. If the resource
+    /// is already in the target state, the method will return immediately.<br/>
+    /// If the resource doesn't reach one of the target states before <paramref name="cancellationToken"/> is signalled, this method
+    /// will throw <see cref="OperationCanceledException"/>.
+    /// </remarks>
+    /// <param name="resourceName">The name of the resouce.</param>
+    /// <param name="targetState"></param>
+    /// <param name="cancellationToken">A <see cref="CancellationToken"/> </param>
+    /// <returns>A <see cref="Task"/> representing the wait operation.</returns>
+#pragma warning disable RS0027 // API with optional parameter(s) should have the most parameters amongst its public overloads
+    public Task WaitForResourceAsync(string resourceName, string? targetState = null, CancellationToken? cancellationToken = null)
+#pragma warning restore RS0027
+    {
+        string[] targetStates = !string.IsNullOrEmpty(targetState) ? [targetState] : [KnownResourceStates.Running];
+        return WaitForResourceAsync(resourceName, targetStates, cancellationToken ?? default);
+    }
+
+    /// <summary>
+    /// Waits for a resource to reach one of the specified states. See <see cref="KnownResourceStates"/> for common states.
+    /// </summary>
+    /// <remarks>
+    /// This method returns a task that will complete when the resource reaches one of the specified target states. If the resource
+    /// is already in the target state, the method will return immediately.<br/>
+    /// Note the returned task will not complete if the resource doesn't reach one of the target states.
+    /// Use <see cref="WaitForResourceAsync(string, IEnumerable{string}, CancellationToken)"/> to pass a <see cref="CancellationToken"/>.
+    /// </remarks>
+    /// <param name="resourceName">The name of the resource.</param>
+    /// <param name="targetStates">The set of states to wait for the resource to transition to one of. See <see cref="KnownResourceStates"/> for common states.</param>
+    /// <returns>A <see cref="Task"/> representing the wait operation.</returns>
+    public Task<string> WaitForResourceAsync(string resourceName, IEnumerable<string> targetStates)
```

Comment:
Do you need this method?

Reply:
Yeah without a default it's kinda dangerous :)
I'll remove